geowrangler/vector_zonal_stats.py

```diff
@@ -0,0 +1,167 @@
+      - split off any existing columns named index and aoi_index and drop them from aoi
+      - create a column 'aoi_index' from aoi's index which will be used as grouping key
+    """
+    # prep for spatial join
+    aoi = aoi.copy()
+
+    # handle existing col named 'index'
+
+    aoi_index_data = None
+    if "index" in list(aoi.columns.values):
+        aoi_index_data = aoi["index"]
+        aoi.drop(labels="index", inplace=True, axis=1)
+
+    # handle existing col named 'aoi_index'
+
+    aoi_col_data = None
+    if "aoi_index" in list(aoi.columns.values):
+        aoi_col_data = aoi["aoi_index"]
+        aoi.drop(labels="aoi_index", inplace=True, axis=1)
+
+    # create index col for broadcast to features
+    aoi.reset_index(level=0, inplace=True)
+    aoi.rename(columns={"index": "aoi_index"}, inplace=True)
+    return aoi, aoi_index_data, aoi_col_data
+
+
+# Internal Cell
+
+
+def _aggregate_stats(
+    aoi: gpd.GeoDataFrame,  # Area of interest
+    groups: pd.core.groupby.DataFrameGroupBy,  # Source data aggregated into groups by 'aoi_index'
+    agg: Dict[str, Any],  # An agg spec to be applied
+) -> gpd.GeoDataFrame:
+    """Aggregate groups and compute agg functions in agg['func'] for agg['column'], map them to output columns in agg['column']
+    and merge them back to aoi dataframe
+    """
+    aggregates = groups[agg["column"]].agg(agg["func"])
+    renames = {k: v for k, v in zip(agg["func"], agg["output"])}
+    aggregates.rename(columns=renames, inplace=True)
+    results = aoi.merge(aggregates, how="left", on="aoi_index", suffixes=(None, "_y"))
+
+    # set NAs to 0 if fillna
+    for i, colname in enumerate(agg["output"]):
+        if agg["fillna"][i]:
+            if colname in list(aoi.columns.values):
+                colname = colname + "_y"  # try if merged df has colname + _y
+            if colname in list(results.columns.values):
+                results[colname].fillna(0, inplace=True)
+
+    return results
+
+
+# Cell
+
+
+def create_zonal_stats(
+    aoi: gpd.GeoDataFrame,  # Area of interest for which zonal stats are to be computed for
+    data: gpd.GeoDataFrame,  # Source gdf containing data to compute zonal stats from
+    aggregations: List,  # List of agg specs, with each agg spect applied to a data column
+    overlap_method: str = "intersects",  # spatial predicate to used in spatial join of aoi and data [geopandas.sjoin](https://geopandas.org/en/stable/docs/user_guide/mergingdata.html#binary-predicate-joins) for more details
+    # categorical_column_options: str = None,
+) -> gpd.GeoDataFrame:
+    """
+    Create zonal stats for area of interest from data using aggregration operations on data columns.
+    Returns the same aoi with additional columns containing the computed zonal features.
+    """
+    # make data crs == aoi crs
+    if not data.crs.equals(aoi.crs):
+        data = data.to_crs(aoi.crs)
+
+    # prep for spatial join
+    aoi, aoi_index_data, aoi_col_data = _prep_aoi(aoi)
```

Comment:
I think it would be easier to just name the index column something like
```
__geowrangleer_aoi_ndex
```
and raise an error if for some reason that column exists at this stage rather than keeping track of everything. we can then drop `__geowrangleer_aoi_ndex`  after all the aggregations.

 A possible issue with replacing columns is if for some reason the index is being used as an aggregation column.

Reply:
Its just that right now, this is flexible enough that it doesn't matter what the aoi column names are.  

If we were to implement your suggestion we would be reducing the flexibility of the function for the user to have to not think about aoi's column names.

As for the issue of replacing columns, it actually wouldn't be possible to use index as the aggregation column for the aoi side since the only columns used in the spatial join from the aoi side are `aoi_index` and the geometry.  The merge of the grouped columns (left join) are only keyed on the `aoi_index` . The function guarantees that the number of rows of the aoi are always preserved in the results. I would also have to add handling the output column name. 

If you feel strongly that the code simplification is worth the reduction in flexibility, then I'll implement it.

Reply:
Leaning toward keeping things simple and less flexible. If users request more flexibility, we can address it then. Personally don't want to add complexity unless there are concrete use cases for it.

Reply:
Ok.